### PR TITLE
add kubectl can-i to see if you can perform an action

### PR DESCRIPTION
**Pull request by deads2k**
Adds `kubectl auth can-i <verb> <resource> [<name>]` so that a user can see if they are allowed to perform an action.

@kubernetes/sig-cli-pr-reviews @fabianofranz 

This particular command satisfies the immediate need of knowing if you can perform an action without trying that action.  When using RBAC in a script that is adding permissions, there is a lag between adding the permission and the permission being realized in the RBAC cache.  As a user on the CLI, you almost never see it, but as a script adding a binding and then using that new power, you hit it quite often.

There are natural follow-ons to the same area (hence the `auth` subcommand) to figure out if someone else can perform an action, what actions you can perform in total, and who can perform a given action.  Someone else is an API we have already, what-can-i-do was a proposed API a while back and a very useful one for interfaces, and who-can is common question if someone is administering a namespace.

### CODE REVIEW

File: pkg/kubectl/cmd/auth/auth.go
Comment (juanvallejo) [inline pkg/kubectl/cmd/auth/auth.go]: nit: should this be the current year?

**Comment by juanvallejo** [comment]
LGTM from a cli perspective

**Review by justinsb — CHANGES_REQUESTED** [review]
I believe the nuances of authorization are much more complicated.  What about quotas, or per-namespace rules, or value-dependent rules?  What about objects that can only be created if an external resource exists?

Can you link to the proposal/design doc for this?

**Comment by deads2k** [comment]
> I believe the nuances of authorization are much more complicated. What about quotas, or per-namespace rules, or value-dependent rules? What about objects that can only be created if an external resource exists?

Is this related to the subcommand I chose (`kubectl auth`) or the general principle of knowing if you're going to get a 403 from our authorization layer (as distinct from admission).  This is different from a "dry-run" using an API object.

**Comment by shiywang** [comment]
@deads2k do we have to add some unit test for this new command ?

**Comment by justinsb** [comment]
The question in my mind is the use-case:  Is this for developers or end users?  I presume it's end users, so why does an end user run the command?

I think I'm just missing some context - is there a feature / issue / similar?

File: hack/make-rules/test-cmd-util.sh
Comment (soltysh) [inline hack/make-rules/test-cmd-util.sh]: Nit: can we have negative test as well?

**Review by soltysh — COMMENTED** [review]
LGTM, except for one nit.

**Comment by deads2k** [comment]
> The question in my mind is the use-case: Is this for developers or end users? I presume it's end users, so why does an end user run the command?
> 
> I think I'm just missing some context - is there a feature / issue / similar?

I'll update the description and add some more detail here.  For starters, its just nice to get a prediction of whether you can do a thing and there are naturally follow-ons to the same area to figure out if someone else can perform an action, what actions you can perform in total, and who can perform a given action.

This particular command satisfies the immediate need of knowing if you can perform an action without trying that action.  When using RBAC in a script that is adding permissions, there is a lag between adding the permission and the permission being realized in the RBAC cache.  As a user on the CLI, you almost never see it, but as a script adding a binding and then using that new power, you hit it quite often.

**Comment by deads2k** [comment]
@pwittrock thoughts on a subcommand for this?  I chose `kubectl auth` anticipating things like whoami and who-can.

**Comment by kargakis** [comment]
fwiw, in openshift we use policy as the parent command name

**Comment by AdoHe** [comment]
@deads2k @kargakis is this PR some kind of upstream feature?

**Comment by deads2k** [comment]
> @deads2k @kargakis is this PR some kind of upstream feature?

It's a command we know that scripts will need (see https://github.com/kubernetes/kubernetes/pull/41077#issuecomment-278653901 or the PR description) based on experience downstream.

File: pkg/kubectl/cmd/auth/cani.go
Comment (deads2k) [inline pkg/kubectl/cmd/auth/cani.go]: Nothing stops user apiservers from including SAR and not registering it to be aggregated. In fact, I would encourage that

Comment (deads2k) [inline pkg/kubectl/cmd/auth/cani.go]: > IIUC, this will only work for core-apiservers and can't use user-apiservers directly. Is that fine?

Consider cases like a someone evaluating a new API server before turning it one.  He'll proxy to the service and then do things to check functionality and permissions.

File: pkg/kubectl/cmd/auth/auth.go
Comment (deads2k) [inline pkg/kubectl/cmd/auth/auth.go]: > Will the sub commands always be about inspection? What about setting / creating things?

Because kube isn't opinionated about authentication and authorization, its very difficult (maybe impossible) to provide generic mutation functionality.  We can provide generic inspection since it mirrors API server needs, but not generic mutation.

File: pkg/kubectl/cmd/auth/cani.go
Comment (deads2k) [inline pkg/kubectl/cmd/auth/cani.go]: > Do we plan to support these as well?

Good point.  I'm ok updating to use that instead.



> What does this do with subcommands?

This command isn't checking for our CLI, its check for our API.  In the example you just used, it would currently create `tls` as a name.  After I update, it will error on invalid input.

**Comment by fabianofranz** [comment]
@deads2k worth adding a release note.  ;)

File: pkg/kubectl/cmd/auth/cani.go
Comment (fabianofranz) [inline pkg/kubectl/cmd/auth/cani.go]: 2017

File: pkg/kubectl/cmd/auth/cani.go
Comment (fabianofranz) [inline pkg/kubectl/cmd/auth/cani.go]: It's worth mentioning in the long desc that `VERB` refer to the API. And then add a few examples for common checks.

File: pkg/kubectl/cmd/auth/cani.go
Comment (fabianofranz) [inline pkg/kubectl/cmd/auth/cani.go]: :+1:

File: pkg/kubectl/cmd/auth/cani.go
Comment (pwittrock) [inline pkg/kubectl/cmd/auth/cani.go]: Yes, the long description should list the commands that are supported

**Comment by deads2k** [comment]
Comments addressed in another commit, will squash before merge.

@pwittrock @AdoHe @fabianofranz How do you feel about the naming of the command?

**Comment by justinsb** [comment]
It seems that the authz API does not match our pattern elsewhere - in the k8s API, we generally communicate status of async operations by updating a field in spec or status.  Is there some reason we can't follow the pattern here?

I suspect the reason is because that concept is ill-defined for HA apiservers, but then how can this operation work?  How do we know the authz operation has completed on all servers, and we haven't just "got lucky" on a particular server?

If we made the authz API work through status, I would be in favor of a `kubectl wait` operation (for example), which could waits for a particular field to be a certain value.  That would be generally applicable, and would probably warrant a top-level command.

If there are scripts that need this, maybe we can compromise by putting this command somewhere where we can deprecate it in future?  `kubectl alpha auth` ?

**Comment by deads2k** [comment]
> It seems that the authz API does not match our pattern elsewhere - in the k8s API, we generally communicate status of async operations by updating a field in spec or status. Is there some reason we can't follow the pattern here?

We made that choice back in 1.1 or 1.2.  It's a non-mutating POST to provide data that reflects the status of "now".  It works quite well for all the delegated authentication and authorization use-cases we've found.

**Comment by deads2k** [comment]
> If there are scripts that need this, maybe we can compromise by putting this command somewhere where we can deprecate it in future? kubectl alpha auth ?

There are clear CLI use-cases for authentication and authorization inspection commands.  While I'm not strictly against an alpha or experimental area, I haven't heard doubt about general utility or suitability.  This is using a GA API that's been vetted for over two years.

**Comment by justinsb** [comment]
I guess I still simply don't understand the problem we're solving.  I agree that inspection of auth results is useful, and I would like to see those commands.

But this PR was introduced as solving "When using RBAC in a script that is adding permissions, there is a lag between adding the permission and the permission being realized in the RBAC cache.", but this command does not solve that problem, particularly in HA scenarios.

If it's actually a user-facing command, perhaps `can-i` implies more than we intend.  We are really querying "am I authorized".  i.e. perhaps if we expressed this as a query on permissions, this would both more correctly express what it does, but also allow for future evolution as well (e.g. maybe we authorize conditionally on IP range, or only if authenticated with two-factor or something like that).

We're firmly into CLI design at this point, so I'll leave this up to the sig-cli triumvirate.

**Comment by deads2k** [comment]
> But this PR was introduced as solving "When using RBAC in a script that is adding permissions, there is a lag between adding the permission and the permission being realized in the RBAC cache.", but this command does not solve that problem, particularly in HA scenarios.

This is no different than the basic `GET?watch=true` which hits a watch cache on an HA server which is not guaranteed to be synced amongst them.  Sometimes it might work, sometimes your resource version will be out of range depending on which API server you hit.

**Comment by deads2k** [comment]
@k8s-bot unit test this
@k8s-bot kubemark e2e test this

File: pkg/kubectl/cmd/auth/cani.go
Comment (pwittrock) [inline pkg/kubectl/cmd/auth/cani.go]: Is this standard syntax to have kind.group?  Do we do group.kind anywhere?

Comment (deads2k) [inline pkg/kubectl/cmd/auth/cani.go]: > Is this standard syntax to have kind.group? Do we do group.kind anywhere?

`kubectl get, patch, update, delete, apply` and everywhere that uses a resource builder uses `resource.group` that matches this help.

Comment (pwittrock) [inline pkg/kubectl/cmd/auth/cani.go]: Very interesting.  This is the first time I have noticed it in the help documentation.

File: pkg/kubectl/cmd/auth/cani_test.go
Comment (pwittrock) [inline pkg/kubectl/cmd/auth/cani_test.go]: Would you add tests for error conditions and all namespaces flag?  I guess testing out the quiet mode would be a bit of a pain since you would need to mock out the exit command.

File: pkg/kubectl/cmd/auth/cani.go
Comment (pwittrock) [inline pkg/kubectl/cmd/auth/cani.go]: nit: Should we provide a user friendly message in the event the api is not enabled?

**Comment by pwittrock** [comment]
lgtm.  Would you add the additional tests suggested and I will formalize it?

File: pkg/kubectl/cmd/auth/cani.go
Comment (deads2k) [inline pkg/kubectl/cmd/auth/cani.go]: > nit: Should we provide a user friendly message in the event the api is not enabled?

This would return our 404, which gets processed through our normal `checkerr` path and should come back looking reasonable to a user.  A general prereq discovery pattern for kubectl would be neat though.  Especially as we get more API servers.

**Comment by deads2k** [comment]
> lgtm. Would you add the additional tests suggested and I will formalize it?

tests added.

**Comment by deads2k** [comment]
@k8s-bot bazel test this
@k8s-bot cross build this

**Comment by deads2k** [comment]
@k8s-bot cvm gke e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot kubemark e2e test this

**Pull request merged**
